src/main/java/teammates/ui/webapi/DeadlineExtensionsWorkerAction.java

```diff
@@ -0,0 +1,156 @@
+package teammates.ui.webapi;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import teammates.common.datatransfer.attributes.CourseAttributes;
+import teammates.common.datatransfer.attributes.DeadlineExtensionAttributes;
+import teammates.common.datatransfer.attributes.FeedbackSessionAttributes;
+import teammates.common.exception.EntityAlreadyExistsException;
+import teammates.common.exception.EntityDoesNotExistException;
+import teammates.common.exception.InvalidParametersException;
+import teammates.common.util.EmailType;
+import teammates.common.util.EmailWrapper;
+import teammates.common.util.Logger;
+import teammates.ui.request.DeadlineExtensionRequest;
+import teammates.ui.request.InvalidHttpRequestBodyException;
+
+/**
+ * Task queue worker action: creates, updates or deletes deadline extensions and sends out appropriate emails.
+ */
+class DeadlineExtensionsWorkerAction extends AdminOnlyAction {
+
+    private static final Logger log = Logger.getLogger();
+
+    @Override
+    public JsonResult execute() throws InvalidHttpRequestBodyException {
+        DeadlineExtensionRequest deadlineExtensionRequest = getAndValidateRequestBody(DeadlineExtensionRequest.class);
+        CourseAttributes course = logic.getCourse(deadlineExtensionRequest.getCourseId());
+        FeedbackSessionAttributes session = logic.getFeedbackSession(deadlineExtensionRequest.getFeedbackSessionName(),
+                deadlineExtensionRequest.getCourseId());
+
+        if (course == null || session == null) {
+            log.severe("Course with id: " + deadlineExtensionRequest.getCourseId() + " or feedback session: "
+                    + deadlineExtensionRequest.getFeedbackSessionName() + " not found when updating deadline extensions");
+            return new JsonResult("Failure");
+        }
+
+        boolean notifyUsers = deadlineExtensionRequest.getNotifyUsers();
+
+        createOrUpdateDeadlineExtensions(course, session,
+                deadlineExtensionRequest.getStudentExtensionsToModify(), false, notifyUsers);
+        createOrUpdateDeadlineExtensions(course, session,
+                deadlineExtensionRequest.getInstructorExtensionsToModify(), true, notifyUsers);
+        revokeDeadlineExtensions(course, session,
+                deadlineExtensionRequest.getStudentExtensionsToRevoke(), false, notifyUsers);
+        revokeDeadlineExtensions(course, session,
+                deadlineExtensionRequest.getInstructorExtensionsToRevoke(), true, notifyUsers);
```

Comment:
I'm not really familiar with task queues, but was wondering about the rationale for both handling deadline extension entities and sending emails under this same worker action - are there benefits to doing this, as compared to say directly creating/updating/deleting deadline extensions when we update a feedback session, and delegating only the part on sending emails to the task queue?

Reply:
Thanks for the review! Rationale here is that it might take awhile to update/create the entities, so if we delegate it to a task queue the user doesn't have to wait for all entities to be created/updated. We can do this because the DeadlineExtension entity is not the source of truth (the `studentDeadlines` and `instructorDeadlines` fields are already updated, and these will be used as the single source of truth).

Reply:
Makes sense, let's stick with this then!